rust/lance/src/dataset/write.rs

```diff
@@ -644,6 +647,103 @@ async fn resolve_commit_handler(
     }
 }
 
+/// Create an iterator of record batch streams from the given source.
+///
+/// If `enable_retries` is true, the source will be collected, and the
+/// resulting iterator will be an infinite iterator that returns the same
+/// batches over and over again. Otherwise, the source will be yielded by a
+/// one-item iterator.
+///
+/// This is used to support retries on write operations.
+async fn new_source_iter(
+    source: SendableRecordBatchStream,
+    enable_retries: bool,
+) -> Result<Box<dyn Iterator<Item = SendableRecordBatchStream> + Send + 'static>> {
+    if enable_retries {
+        let schema = source.schema();
+
+        // If size hint shows there is only one batch, spilling has no benefit, just keep that
+        // in memory. (This is a pretty common case.)
+        let size_hint = source.size_hint();
+        if size_hint.0 == 1 && size_hint.1 == Some(1) {
+            let batches: Vec<RecordBatch> = source.try_collect().await?;
+            Ok(Box::new(std::iter::repeat_with(move || {
+                Box::pin(RecordBatchStreamAdapter::new(
+                    schema.clone(),
+                    futures::stream::iter(batches.clone().into_iter().map(Ok)),
+                )) as SendableRecordBatchStream
+            })))
+        } else {
+            // TODO: allow buffering up to 100MB in memory before spilling to disk.
```

Comment:
Does this not do any in-memory buffering at all right now? 100MB seems small for some server-side use-cases, and we should make it configurable.

Reply:
It doesn't do any in-memory buffering, no.

Reply:
> we may want to buffer more than 100MB in memory

I was wondering more if you think it's dangerous to ship as-is. The only times I can think where you might want to change this seem to be edge cases:

1. You are writing 1GB of data, but the latency hit from writing to disk is meaningful. If the final destination is object storage, and your cache is an SSD or even just HDD, I don't think this is true. Could be true is final destination is also SSD.
2. You are writing 10GB of data, and you have 10GB of memory available but not 10GB of disk available. I think 99% of times a computer will have more disk than memory available.

Are there others you have in mind?